src/sentry/release_health/metrics_sessions_v2.py

```diff
@@ -0,0 +1,607 @@
+            errored = max(0, all_errored - abnormal - crashed)
+
+            return [
+                _StatusValue("abnormal", abnormal),
+                _StatusValue("crashed", crashed),
+                _StatusValue("errored", errored),
+                _StatusValue("healthy", healthy),
+            ]
+
+        # The "init" branch generates values for all statuses, so nothing
+        # to do here:
+        return []
+
+
+class _SessionDurationField(_OutputField):
+    def __init__(
+        self, name: SessionsQueryFunction, column: _VirtualColumnName, group_by_status: bool
+    ) -> None:
+        self._name = name
+        self._column = column
+
+        self.group_by_status = group_by_status
+
+    def get_name(self) -> SessionsQueryFunction:
+        return self._name
+
+    def get_values(self, data_points: _DataPoints, key: _DataPointKey) -> Sequence[_StatusValue]:
+        assert key.raw_session_status is None, key  # session.duration does not have status tag
+        value = 1000.0 * data_points[key]  # sessions backend stores milliseconds
+        if self.group_by_status:
+            # Only 'healthy' sessions have duration data:
+            return [
+                _StatusValue("abnormal", None),
+                _StatusValue("crashed", None),
+                _StatusValue("errored", None),
+                _StatusValue("healthy", value),
+            ]
+        else:
+            return [
+                _StatusValue(None, value),
+            ]
+
+
+def _fetch_data(
+    org_id: int,
+    query: QueryDefinition,
+) -> Tuple[_SnubaDataByMetric, Mapping[Tuple[_MetricName, _VirtualColumnName], _OutputField]]:
+    """Build & run necessary snuba queries"""
+    conditions = [
+        Condition(Column("org_id"), Op.EQ, org_id),
+        Condition(Column("project_id"), Op.IN, query.filter_keys["project_id"]),
+        Condition(Column(TS_COL_QUERY), Op.GTE, query.start),
+        Condition(Column(TS_COL_QUERY), Op.LT, query.end),
+    ]
+
+    conditions.extend(_get_filter_conditions(org_id, query.conditions))
+
+    # It greatly simplifies code if we just assume that these two tags exist:
+    # TODO: Can we get away with that assumption?
+    tag_key_session_status = _resolve_ensured(org_id, "session.status")
+
+    groupby_tags = [field for field in query.raw_groupby if field != "project"]
+
+    tag_keys = {field: _resolve(org_id, field) for field in groupby_tags}
+    groupby = {
+        field: Column(f"tags[{tag_id}]")
+        for field, tag_id in tag_keys.items()
+        if tag_id is not None  # exclude unresolved keys from groupby
+    }
+
+    if "project" in query.raw_groupby:
+        groupby["project"] = Column("project_id")
+
+    data: List[Tuple[_MetricName, _SnubaData]] = []
+
+    #: Find the field that needs a specific column in a specific metric
+    metric_to_output_field: MutableMapping[
+        Tuple[_MetricName, _VirtualColumnName], _OutputField
+    ] = {}
+
+    def get_query(
```

Comment:
tbh I would move atleast all these long functions outside of this main function.. imo will make it much easier to read cause there is already a lot going on in this function

Reply:
Moved these two helper functions out to the module level now. In order to keep their signature simple, there is some code that now runs for every call, which means there are more calls to the indexer service. This should not be a problem as soon as there is an indexer implementation with a built-in in-memory cache.